weave/trace_server/refs_internal.py

```diff
@@ -17,16 +17,82 @@
 OBJECT_ATTR_EDGE_NAME = "attr"
 TABLE_ROW_ID_EDGE_NAME = "id"
 
+valid_edge_names = (
+    DICT_KEY_EDGE_NAME,
+    LIST_INDEX_EDGE_NAME,
+    OBJECT_ATTR_EDGE_NAME,
+    TABLE_ROW_ID_EDGE_NAME,
+)
+
 
 class InvalidInternalRef(ValueError):
     pass
 
 
+def quote_select(s: str, quote_chars: tuple[str, ...] = ("/", ":", "%")) -> str:
+    """We don't need to quote every single character, rather
+    we just need to quote the characters that are used as
+    delimiters in the URI. Right now, we only use "/", ":", and "%". Moreover,
+    the only user-controlled fields are the object name and the extra fields.
+
+    All other fields are generated by the system and are safe for parsing.
+    """
+    # We have to be careful here since quoting creates new "%" characters.
+    # We don't want to double-quote characters.
+    if "%" in quote_chars and quote_chars[-1] != "%":
+        raise ValueError("Quoting '%' must be the last character in the list.")
+
+    for c in quote_chars:
+        s = s.replace(c, urllib.parse.quote(c))
+
+    return s
+
+
+def validate_extra(extra: list[str]) -> None:
+    if len(extra) % 2 != 0:
+        raise InvalidInternalRef("Extra fields must be key-value pairs.")
+    for i, e in enumerate(extra):
+        if i % 2 == 0:
+            if e not in (
+                DICT_KEY_EDGE_NAME,
+                LIST_INDEX_EDGE_NAME,
+                OBJECT_ATTR_EDGE_NAME,
+                TABLE_ROW_ID_EDGE_NAME,
+            ):
+                raise InvalidInternalRef(
+                    f"Invalid extra edge name at index {i}: {extra}"
+                )
+        else:
+            # There is only a single rule here:
+            if extra[i - 1] == LIST_INDEX_EDGE_NAME:
+                try:
+                    int(e)
+                except ValueError:
+                    raise InvalidInternalRef(
+                        f"Invalid list edge value at index {i}: {extra}"
+                    )
+            pass
+
+
+def validate_no_slashes(s: str, field_name: str) -> None:
+    if "/" in s:
+        raise InvalidInternalRef(f"{field_name} cannot contain '/'")
+
+
+def validate_no_colons(s: str, field_name: str) -> None:
+    if ":" in s:
+        raise InvalidInternalRef(f"{field_name} cannot contain ':'")
+
+
 @dataclasses.dataclass
 class InternalTableRef:
     project_id: str
     digest: str
 
+    def __post_init__(self) -> None:
+        validate_no_slashes(self.project_id, "project_id")
+        validate_no_slashes(self.digest, "digest")
```

Comment:
While these will guarantee no slash at init time, they won't if the user changes these fields after.

I think in general refs should be immutable, so we can just set `@dataclass(frozen=True)`

Otherwise, if refs are mutable, then we need to use a property system like:

```py
@dataclass
class InternalTableRef:
    _project_id: field(init=False, repr=False)
    
    def __post_init__(self):
        self.project_id = self._project_id
        
    @property
    def project_id(self):
        return self._project_id
    
    @project_id.setter
    def project_id(self, value):
        validate_no_slashes(value, "project_id")
        self._project_id = value
```

Reply:
great call - we should make these frozen - will do that now.